### PR TITLE
[Docs] add code review guidelines

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Description

This document was originally exported from my notes into [this notion doc](https://www.notion.so/pocketnetwork/Code-Review-Guidelines-proposal-5e22dbb0d40d44f69b5ccab42f64bfe5). I started it a while back and it represents a collaborative effort across multiple teams I've been a part of. This PR incorporates all the immediately actionable feedback in an attempt to get an initial, Pocket specific, version in the repo and move the conversation/iteration to GitHub.

### Summary generated by Reviewpad on 20 Jul 23 00:14 UTC
This pull request includes the addition of a new file called CODE_REVIEW_GUIDELINES.md in the docs/development directory. It outlines code quality and code review guidelines, including best practices and expectations for pull request authors and reviewers. It also provides information on review comments, finishing a review, and merging. Additionally, references to the CODE_REVIEW_GUIDELINES.md file were added in the README.md file. Two new images were also added in the assets directory.

## Issue

N/A

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [x] Documentation
- [ ] Other 

## List of changes

- Adds code review guidelines docs
- Adds code review guidelines doc links to main README

## Testing

- [ ] `make develop_test`; if any code changes were made
- [ ] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [ ] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [ ] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [ ] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [ ] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

**Comment by Olshansk** [comment]
Also adding @adshmh to the review as he is just joining the team.

File: docs/development/CODE_REVIEW_GUIDELINES.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+It also encapsulates ("rough consensus") that we may not always precisely agree and that's okay.
+
+## Code Reviews
+
+One tactic often employed to produce "maintainable" or "more-readable" code is a "code review".
+These can take many shapes and forms and often have goals beyond simply "code quality".
+
+Broadly speaking, code reviews involve developers looking at some proposed new code (or code changes).
+This is *usually* developers other than the author (although a "self-review" can also be employed).
+In many projects, such attention is a scarce commodity: most programmers would rather write code than read someone else's.
```
Comment (Olshansk) [inline docs/development/CODE_REVIEW_GUIDELINES.md:37]: Writing code is the "easy & fun" thing to do ;)

File: docs/development/CODE_REVIEW_GUIDELINES.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+## Code Review Guidelines
+
+All participants must adhere to the overall idea that the review is an attempt to achieve "better" code.
+This is a vague statement on purpose.
+Participants must be cautious in their criticism and generous with praise.
```
Comment (Olshansk) [inline docs/development/CODE_REVIEW_GUIDELINES.md:43]: Thanks Bryan!

File: docs/development/CODE_REVIEW_GUIDELINES.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+If the commits be (re)organized (i.e. reordered and/or amended) such that there is a commit at which the tests are passing prior to the conclusion of the main change, that's a signal that there's likely a logic split which can be made at that point in such a (re)arrangement.
+
+Use the following guidelines to evaluate whether a Pull Request should be approved:
+
+- *Every* Pull Request should have tests or justification otherwise; esp. bug fixes.
+- *Every* Pull Request should have at least 1 approval from a team member internal or external to the project. Exceptions made by repository maintainer(s), as necessary, on a case-by-case basis.
+
+During review, submit feedback using line comments (A); prefer "Add\[ing a\] single comment" over "Start[ing] a review" (B).
+
+*(Once a review has been started, the option to add single comments is removed. Preferring single comments allows feedback to happen even in the event of an interrupted review.)*
```
Comment (Olshansk) [inline docs/development/CODE_REVIEW_GUIDELINES.md:73]: I personally prefer Starting reviews because I often go back and edit/remove comments if my question/comment was addressed elsewhere.

Is this what you found yourself doing recently?

File: docs/development/CODE_REVIEW_GUIDELINES.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+### A
```
Comment (Olshansk) [inline docs/development/CODE_REVIEW_GUIDELINES.md:75]: These A-B-C section headers (which are not in the ToC) are a bit confusing. What is the intention here?

File: docs/development/CODE_REVIEW_GUIDELINES.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+![github_line_comment.png](./assets/github_line_comment.png)
+
+### B
+
+![line_comment_dialog_start.png](./assets/line_comment_dialog_start.png)
+
+**Referencing Issues Across Repositories:**
```
Comment (Olshansk) [inline docs/development/CODE_REVIEW_GUIDELINES.md:83]: Can you add a section on referencing issues/PRs from the same repository using simply `#issue`?

File: docs/development/CODE_REVIEW_GUIDELINES.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+- Core member needs to approve PR
+- Author should merge in the PR themselves (following instructions above)
+
+**Authors are non-regular external contributors:**
+
+- Core member needs to approve PR
+- Core member can merge PR on behalf of contributor  (following instructions above)
+
+**Pull Request Authors:**
+Specify whether or not the Pull Request can be merged by the last approving reviewer by use of (or lack thereof) the "Reviewer can merge" label. If no such label exists for a given project, create it.
```
Comment (Olshansk) [inline docs/development/CODE_REVIEW_GUIDELINES.md:118]: Is this `Reviewer can merge` label a new idea? Can you add it if this is the pattern we'll use?

File: docs/development/CODE_REVIEW_GUIDELINES.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+**Reviewers:**
+If you would give the last approving review **and** the Pull Request has the "Reviewer can merge" label, **merge**.
```
Comment (Olshansk) [inline docs/development/CODE_REVIEW_GUIDELINES.md:121]: Can you please add 2 things to this document:

1. What we discussed in last monday's sync around draft PRs that need preliminary feedback but not ready to merge.

3. Asking everyone to fill out the metadata for the RPs per the discussion here: https://discord.com/channels/824324475256438814/1003704237324259418/1110608286664892498

Screenshot 2023-05-23 at 3 42 20 PM

**Review by Olshansk — CHANGES_REQUESTED** [review]
@bryanchriswhite I know I create a little bit of extra work for you with the comments but really appreciate you putting this together. It'll make it much easier to just have a single place to point to!

**Comment by bryanchriswhite** [comment]
(note to self: add a bit about preferring to link to PR comments from the "conversation" tab because linking via "files changed" can easily produce broken links.)

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 3 above)
Comment (adshmh) [reply docs/development/CODE_REVIEW_GUIDELINES.md:73]: > I personally prefer Starting reviews because I often go back and edit/remove comments if my question/comment was addressed elsewhere.
> 
Same here. Review comments stay pending until you finish the review, so I can leave placeholder comments on items that may need a change request. It is very common to encounter something else further down the diff that establishes a previously suspicious item as correct.

File: docs/development/CODE_REVIEW_GUIDELINES.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+Confirm that all items in the "required checklist" are checked or not applicable.
+
+If you believe the Pull Request looks good to merge, select "Approve" and submit your review (C).
+
+### C
+
+![submit_review.png](./assets/submit_review.png)
+
+### Merging
```
[the remark this answers is not in the document]
Comment (adshmh) [reply docs/development/CODE_REVIEW_GUIDELINES.md:102]: Also, when merging, some rebasing / force pushing may be necessary: e.g. when CHANGELOG files have been modified by other merged PRs, which will block merge commits due to the commit hooks failure, and leave rebase as the only option (unless I am overlooking some git feature). Some hints regarding such scenarios might be helpful.

[the remark this answers is not in the document]
Comment (Olshansk) [reply docs/development/CODE_REVIEW_GUIDELINES.md:102]: Because we "squash & merge" our PRs to main (since some PRs might end up having dozens of commits), the simplest thing to do is to simply do a `git merge main` rather than a `git rebase main`. It saves times and avoids needing to force push.

[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply docs/development/CODE_REVIEW_GUIDELINES.md:102]: > I suggest we add a section: `Addressing Reviews` to help with a few items, e.g.:
> 
> * Do not rebase (and subsequently push -f) the PR branch, as it would make it challenging for the reviewer to identify what has changed between their review and the latest PR update. Instead push more commits to address review comments. The squash merge method will combine all the PR commits into one.

I have a strong opinion about **supporting rebase workflows** to whatever extent possible as I believe it can add significant value by improving readability of PRs on multiple levels. I've been developing and using a workflow I've pet-named "quantum git theory" which preserves the sequence of events on the PR timeline with respect to human interactions / "observations" (the git-time continuum). 

TLDR; contributors can rewrite history **up to the most recent human interaction** (e.g.: code comment, review, commit). If done correctly, this has the effect of **preserving the order of human-backed PR timeline events** with respect to commits (which I think is the main concern).

As @Olshansk points out, the consequences of what we do in version control on any given PR are constrained to that context of that branch due to the squash merge with main.

Perhaps we can discuss it further in a protocol hour. I have some other git-related things that might be useful to share with the team; perhaps, vice-versa.

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 2 above)
Comment (bryanchriswhite) [reply docs/development/CODE_REVIEW_GUIDELINES.md:43]: I see what you did there :wink:

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 4 above)
Comment (bryanchriswhite) [reply docs/development/CODE_REVIEW_GUIDELINES.md:75]: The intention was for those headers to act as labels for the respective images which they contain. The A, B, C labels are referenced in the preceding paragraphs like so: "(A)"

Do you have any suggestions for how to better communicate that intention and/or organize things here?

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 5 above)
Comment (bryanchriswhite) [reply docs/development/CODE_REVIEW_GUIDELINES.md:83]: Of course!

(note to self: you get a neat status icon if you put the link in an (un/)ordrered list):

### Example

#100 

VS

- #100 

OR

1. #100

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 3 above)
Comment (bryanchriswhite) [reply docs/development/CODE_REVIEW_GUIDELINES.md:73]: > Is this what you found yourself doing recently?

I've been doing it this way for several years now. My experience has been that I may not have time to run down every thread in one review session but more often than not review comments / suggestions are independently actionable and non-blocking with respect to one another. The intention was to increase team velocity by not withholding already actionable feedback.

I acknowledge and can also relate the case where you aren't sure about a comment just yet and want to wait until you're sure you've fully groked it or ran down / validated a related concern. I've chosen to prefer editing a premature comment over having more than a couple of comments pending at once. With only a couple of comments pending, I've found it pretty reasonable to leave the text in the unsubmitted textarea and/or copy things between there and my notes for larger scope feedback / investigations.

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply docs/development/CODE_REVIEW_GUIDELINES.md:102]: > Some hints regarding such scenarios might be helpful.

Agreed! :raised_hands: 

> Also, when merging, some rebasing / force pushing may be necessary: e.g. when CHANGELOG files have been modified by other merged PRs, which will block merge commits due to the commit hooks failure, and leave rebase as the only option (unless I am overlooking some git feature). 

AFAIK merging should never necessitate force pushing on its own. When merging an upstream branch which contains a conflict you can either resolve the conflict *in the merge commit* OR "accept theirs" (remote) and re-apply your changes in a subsequent commit. In either case, unless you changed history prior to the merge, it should be a fast-forward from the remote HEAD to your local HEAD (post-merge, etc.).

Here's an excerpt from #461 > "testing methodology" > "merge conflict test" which  illustrates the simplest variation of scenario:

  ```mermaid
  gitGraph
      commit
      branch FeatureA
      commit id: "[v0.0.0.1 - 2023-04-03]" type: HIGHLIGHT
      checkout main
      commit id: "[v0.0.0.1 - 2023-04-01]" type: HIGHLIGHT
      commit id: "[v0.0.0.2 - 2023-04-05]" type: HIGHLIGHT
      checkout FeatureA
      merge main
  ```

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 6 above)
Comment (bryanchriswhite) [reply docs/development/CODE_REVIEW_GUIDELINES.md:118]: New to Pocket, yes. Unless there are any principled objections, I would like for us to give it a try.

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 7 above)
Comment (dylanlott) [reply docs/development/CODE_REVIEW_GUIDELINES.md:121]: For me personally I didn't realize how to add Metadata from the Issue or PR view, so maybe this should have a section about how to fill that out i.e. what epics we use, what sizes mean, etc... with a screenshot accompanying.

File: docs/development/CODE_REVIEW_GUIDELINES.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+Selecting the "*best*" solution is often a matter of style.
+Sometimes the best solution is one that fits the surrounding code in the most cohesive way.
+Terms like "maintainability" or "readability" are used; these address the ability of other contributors to understand and improve the code.
+
+Unlike correctness or performance concepts, there's no single metric or mathematical solution that can be optimized to achieve better code quality.
+Thus, we rely on human judgement.
+
+In this document, we outline a framework to guide that human judgement towards -- collectively -- "better code".
+
+For decades, the IETF (Internet Engineering Task Force) has used the motto "rough consensus and running code".
```
[the remark this answers is not in the document]
Comment (dylanlott) [reply docs/development/CODE_REVIEW_GUIDELINES.md:26]: I didn't know this either but I really like this definition! Can we link to that IETF document here? That would be useful reading I think.

**Review by dylanlott — APPROVED** [review]
Looks good!

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply docs/development/CODE_REVIEW_GUIDELINES.md:26]:
```suggestion
For decades, the IETF (Internet Engineering Task Force) has used the motto ["rough consensus and running code"](https://www.ietf.org/runningcode/).
```
---

image

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 3 above)
Comment (bryanchriswhite) [reply docs/development/CODE_REVIEW_GUIDELINES.md:73]: I'm realizing a couple of other hangups that I have with the combined review comment UX:

1. It's likely to cause duplicate comments in the PR timeline
2. The review comment itself and related comments can't be "resolved" (only supported action is limite emoji reaction)

I threw together a quick [demo PR on my fork](https://github.com/bryanchriswhite/pocket/pull/8) where you can see that the third reviews comments are duplicated and unresolvable in the combined review copy.

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 1 above)
Comment (Olshansk) [reply docs/development/CODE_REVIEW_GUIDELINES.md:37]: Yea, let's do it! Future GPT versions will use it in their training

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 6 above)
Comment (Olshansk) [reply docs/development/CODE_REVIEW_GUIDELINES.md:118]: No objections, let's do it.

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 7 above)
Comment (Olshansk) [reply docs/development/CODE_REVIEW_GUIDELINES.md:121]: @adshmh Can you open a ticket for it?

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 5 above)
Comment (Olshansk) [reply docs/development/CODE_REVIEW_GUIDELINES.md:83]: That is a fantastic PRO tip!

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 4 above)
Comment (Olshansk) [reply docs/development/CODE_REVIEW_GUIDELINES.md:75]: The heading can be `Figure A: XXX` or `Figure 1: XXX` and you can reference it accordingly.

**Review by Olshansk — CHANGES_REQUESTED** [review]

**Comment by Olshansk** [comment]
@bryanchriswhite Bump on hopefully getting this over the finish line and iterating on main

File: docs/development/CODE_REVIEW_GUIDELINES.md (hunk 7 above)
Comment (adshmh) [reply docs/development/CODE_REVIEW_GUIDELINES.md:135]: > @adshmh Can you open a ticket for it?

Done: https://github.com/pokt-network/pocket/issues/864

**Comment by Olshansk** [comment]
@bryanchriswhite Do you think you'll be able to finish this off before EOW?

**Review by Olshansk — APPROVED** [review]
Made some minor edits but going to approve & merge in so we can iterate on top of it.
